src/librustc_mir/build/mod.rs

```diff
@@ -872,6 +875,23 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
         self.into(&Place::RETURN_PLACE, block, body)
     }
 
+    fn set_correct_source_scope_for_arg(
+        &mut self,
+        arg_hir_id: hir::HirId,
+        pattern_span: Span
+    ) {
+        let tcx = self.hir.tcx();
+        let root = tcx.maybe_lint_level_root_bounded(
+            arg_hir_id,
+            self.hir.root_lint_level
+        );
+        self.source_scope = self.new_source_scope(
```

Comment:
A new source scope only needs to be created if `root != self.hir.root_lint_scope`